src/copy.rs

```diff
@@ -239,38 +310,18 @@ impl<R: AsyncRead> BufReader<R> {
 }
 
 impl<R: AsyncRead> ResizeBufRead for BufReader<R> {
-    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<&[u8]>> {
+    fn poll_bytes(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<Bytes>> {
         let me = self.project();
 
-        // If we've reached the end of our internal buffer then we need to fetch
-        // some more data from the underlying reader.
-        // Branch using `>=` instead of the more correct `==`
-        // to tell the compiler that the pos..cap slice is always valid.
-        if *me.pos >= *me.cap {
-            debug_assert!(*me.pos == *me.cap);
-            let mut buf = tokio::io::ReadBuf::new(me.buf);
-            ready!(me.inner.poll_read(cx, &mut buf))?;
-            *me.cap = buf.filled().len();
-            *me.pos = 0;
-        }
-        Poll::Ready(Ok(&me.buf[*me.pos..*me.cap]))
-    }
-
-    fn consume(self: Pin<&mut Self>, amt: usize) {
-        let me = self.project();
-        *me.pos = cmp::min(*me.pos + amt, *me.cap);
+        // Give us enough space to read a full chunk
+        me.buf.reserve(*me.buffer_size);
```

Comment:
If the buffer size was increased it looks like allocation would be done here, at reserve during this poll, rather than during the resize. Is that on purpose? I suppose it should mean that the buffer is empty when we allocate to reduce any potential data movement.